# IMATH Exercises - Block 8

## Machine Accuracy

We can compute the machine epsilon $e_{mach}$ for IEEE 754 double precision using the following code

In [58]:
x = 1;
while 1 + x > 1
    x
    x = x/2;
end

x =  1
x =  0.50000
x =  0.25000
x =  0.12500
x =  0.062500
x =  0.031250
x =  0.015625
x =  0.0078125
x =  0.0039062
x =  0.0019531
x =  0.00097656
x =  0.00048828
x =  0.00024414
x =  0.00012207
x =  0.000061035
x =  0.000030518
x =  0.000015259
x =  0.0000076294
x =  0.0000038147
x =  0.0000019073
x =  0.00000095367
x =  0.00000047684
x =  0.00000023842
x =  0.00000011921
x =  0.000000059605
x =  0.000000029802
x =  0.000000014901
x =  0.0000000074506
x =  0.0000000037253
x =  0.0000000018626
x =  0.00000000093132
x =  0.00000000046566
x =  0.00000000023283
x =  0.00000000011642
x =  0.000000000058208
x =  0.000000000029104
x =  0.000000000014552
x =    7.2760e-12
x =    3.6380e-12
x =    1.8190e-12
x =    9.0949e-13
x =    4.5475e-13
x =    2.2737e-13
x =    1.1369e-13
x =    5.6843e-14
x =    2.8422e-14
x =    1.4211e-14
x =    7.1054e-15
x =    3.5527e-15
x =    1.7764e-15
x =    8.8818e-16
x =    4.4409e-16
x =    2.2204e-16


We find $e_{mach} = 2.2204 * 10^{-16}$

## Exercise 2 

Plot the error of $1 - ((1+x)-1)/(x-1)$

In [1]:
for i = -7:7
    x = 10^i;
    1-((1+x)-1)/(x-1)
end

ans =  1.0000
ans =  1.0000
ans =  1.0000
ans =  1.0001
ans =  1.0010
ans =  1.0101
ans =  1.1111
ans = -Inf
ans = -0.11111
ans = -0.010101
ans = -0.0010010
ans = -0.00010001
ans = -0.000010000
ans = -0.0000010000
ans = -0.00000010000


## Exercise 4

In [49]:
U_k = 2*sqrt(2);
max_iter = 40;
printf('Double\t\t Float\t\t K\n')


for k=2:max_iter
    U_k1 = 2^k*sqrt(2*(1-sqrt(1-(2^(-k)*U_k)^2)));
    printf('%d\t\t %f\t %d\n',U_k1, U_k1, k)
    U_k = U_k1;
end;


Double		 Float		 K
3.06147		 3.061467	 2
3.12145		 3.121445	 3
3.13655		 3.136548	 4
3.14033		 3.140331	 5
3.14128		 3.141277	 6
3.14151		 3.141514	 7
3.14157		 3.141573	 8
3.14159		 3.141588	 9
3.14159		 3.141591	 10
3.14159		 3.141592	 11
3.14159		 3.141593	 12
3.14159		 3.141593	 13
3.14159		 3.141593	 14
3.14159		 3.141593	 15
3.14159		 3.141593	 16
3.14159		 3.141593	 17
3.14159		 3.141594	 18
3.1416		 3.141597	 19
3.1416		 3.141597	 20
3.14167		 3.141674	 21
3.14183		 3.141830	 22
3.14245		 3.142451	 23
3.14245		 3.142451	 24
3.16228		 3.162278	 25
3.16228		 3.162278	 26
3.4641		 3.464102	 27
4		 4.000000	 28
0		 0.000000	 29
0		 0.000000	 30
0		 0.000000	 31
0		 0.000000	 32
0		 0.000000	 33
0		 0.000000	 34
0		 0.000000	 35
0		 0.000000	 36
0		 0.000000	 37
0		 0.000000	 38
0		 0.000000	 39
0		 0.000000	 40


In [9]:
function x = ComputePi_01

  Uold1 = 2*sqrt(2);
  Uold2 = 2*sqrt(2);
  maxit = 24;

  fprintf('rel. error for bad (left) and good (right) method\n');
  
  for i=2:maxit
    k = i;
    a1 = (Uold1/2^k)^2;
    Unew1 = 2^k*sqrt(2*(1-sqrt(1-a1)));
    Uold1 = Unew1;
    a2 = (Uold2/2^k)^2;
    Unew2 = 2^k*sqrt(2*a2/(1+sqrt(1-a2)));
    Uold2 = Unew2;
    fprintf('%16.14e %16.14e\n',abs(pi-Unew1)/pi,abs(pi-Unew2)/pi);
  end;
end

In [10]:
ComputePi_01

rel. error for bad (left) and good (right) method
2.55046415955671e-02 2.55046415955671e-02
6.41314885579401e-03 6.41314885579401e-03
1.60560696438116e-03 1.60560696438144e-03
4.01546850325250e-04 4.01546850320727e-04
1.00395783863265e-04 1.00395783858035e-04
2.50995129994164e-05 2.50995129497997e-05
6.27491342261881e-06 6.27491367437739e-06
1.56872974177577e-06 1.56873063332059e-06
3.92184886379381e-07 3.92182796684276e-07
9.80326701348349e-08 9.80457077232272e-08
2.45241179447573e-08 2.45114273902202e-08
6.40647820848329e-09 6.12785681221557e-09
3.87636533591722e-10 1.53196420305389e-09
2.63197006556094e-09 3.82991086102969e-10
1.47103964621716e-08 9.57477008467494e-11
8.19170108940481e-08 2.39368545326944e-11
4.68426546739940e-07 5.98410761468422e-12
1.24144517018555e-06 1.49585020618875e-12
1.24144517018555e-06 3.73750514568424e-13
2.59777256198812e-05 9.31549126704203e-14
7.54484510070300e-05 2.28999937065373e-14
2.73306885715936e-04 5.37160346202727e-15
2.73306885715936e-04 9.895

## Exercise 6

In [4]:
function x = GaussElimination(A, b)
    % Solve the linear system Ax = b
    % using Gaussian elimination without pivoting
    % A is an n by n matrix
    % b is an n by k matrix (k copies of n-vectors)
    % x is an n by k matrix (k copies of solution vectors)

    [n, n] = size(A);     % Find size of matrix A
    [n, k] = size(b);     % Find size of matrix b
    x = zeros(n,k);       % Initialize x 
    
    for i = 1:n-1
        m = -A(i+1:n,i)/A(i,i); % multipliers
        A(i+1:n,:) = A(i+1:n,:) + m*A(i,:);
        b(i+1:n,:) = b(i+1:n,:) + m*b(i,:);
    end;

    % Use back substitution to find unknowns
    x(n,:) = b(n,:)/A(n,n);
    
    for i = n-1:-1:1
        x(i,:) = (b(i,:) - A(i,i+1:n)*x(i+1:n,:))/A(i,i);
    end
endfunction

In [5]:
A = [2 -1 3; -4 6 -5; 6 13 16]
b = [2; 1; 3]

result = GaussElimination(A,b)

% Check with A * result - b
check = A * result - b

A =

    2   -1    3
   -4    6   -5
    6   13   16

b =

   2
   1
   3

result =

   14.0833
    3.1667
   -7.6667

check =

   0
   0
   0



In [33]:
function y = poly2(x)
y = sin(x);
end;

[xval, yval, flag] = fminbnd('poly2', 0, pi)





xval =  0.000045994
yval =  0.000045994
flag =  1


In [35]:
x = (pi^5)/(180*(2*10^(-7)))
nthroot(x, 4)

x =  8500546.79959
ans =  53.996


In [37]:
(exp(-2^(-10))+ exp(-2^(-10)))/(2*(2^(-10)))

ans =  1023.0
